File: buggy-ios/buggy-ios/ButtonPanelView.swift
//
//  ButtonPanelView.swift
//  buggy-ios
//

import SwiftUI

struct CalcButton {
    let label: String
    let type: String
    let action: String
    var wide: Bool = false
}

private let buttons: [CalcButton] = [
    CalcButton(label: "MC", type: "memory", action: "memory-clear"),
    CalcButton(label: "MR", type: "memory", action: "memory-recall"),
    CalcButton(label: "M+", type: "memory", action: "memory-add"),
    CalcButton(label: "M\u{2212}", type: "memory", action: "memory-sub"),

    CalcButton(label: "C", type: "clear", action: "clear"),
    CalcButton(label: "\u{232b}", type: "clear", action: "backspace"),
    CalcButton(label: "%", type: "operator", action: "percent"),
    CalcButton(label: "\u{00f7}", type: "operator", action: "/"),

    CalcButton(label: "7", type: "digit", action: "7"),
    CalcButton(label: "8", type: "digit", action: "8"),
    CalcButton(label: "9", type: "digit", action: "9"),
    CalcButton(label: "\u{00d7}", type: "operator", action: "*"),

    CalcButton(label: "4", type: "digit", action: "4"),
    CalcButton(label: "5", type: "digit", action: "5"),
    CalcButton(label: "6", type: "digit", action: "6"),
    CalcButton(label: "\u{2212}", type: "operator", action: "-"),

    CalcButton(label: "1", type: "digit", action: "1"),
    CalcButton(label: "2", type: "digit", action: "2"),
    CalcButton(label: "3", type: "digit", action: "3"),
    CalcButton(label: "+", type: "operator", action: "+"),

    CalcButton(label: "0", type: "digit", action: "0", wide: true),
    CalcButton(label: ".", type: "digit", action: "."),
    CalcButton(label: "=", type: "equals", action: "="),
]

struct ButtonPanelView: View {
    let onButton: (String) -> Void

    private let columns = Array(repeating: GridItem(.flexible(), spacing: 8), count: 4)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 8) {
            ForEach(Array(buttons.enumerated()), id: \.offset) { _, btn in
                if btn.wide {
                    Button(action: { onButton(btn.action) }) {
                        Text(btn.label)
                            .font(.system(size: btn.type == "memory" ? 14 : 18, weight: .medium))
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 16)
                            .background(backgroundColor(for: btn.type))
                            .foregroundColor(foregroundColor(for: btn.type))
                            .cornerRadius(10)
                    }
                    .gridCellColumns(2)
                } else {
                    Button(action: { onButton(btn.action) }) {
                        Text(btn.label)
                            .font(.system(size: btn.type == "memory" ? 14 : 18, weight: .medium))
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, btn.type == "memory" ? 10 : 16)
                            .background(backgroundColor(for: btn.type))
                            .foregroundColor(foregroundColor(for: btn.type))
                            .cornerRadius(10)
                    }
                }
            }
        }
    }

    private func backgroundColor(for type: String) -> Color {
        switch type {
        case "digit": return Color(red: 0.10, green: 0.10, blue: 0.24)
        case "operator": return Color(red: 0.91, green: 0.27, blue: 0.37)
        case "equals": return Color(red: 0.06, green: 0.61, blue: 0.34)
        case "clear": return Color(red: 0.33, green: 0.20, blue: 0.51)
        case "memory": return Color(red: 0.10, green: 0.10, blue: 0.24)
        default: return Color.gray
        }
    }

    private func foregroundColor(for type: String) -> Color {
        switch type {
        case "digit": return Color(red: 0.90, green: 0.90, blue: 0.90)
        case "memory": return Color(red: 0.53, green: 0.57, blue: 0.69)
        default: return .white
        }
    }
}

File: buggy-android/app/src/main/java/com/bughuntbingo/buggyandroid/ui/ButtonPanelView.kt
package com.bughuntbingo.buggyandroid.ui

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class CalcButton(
    val label: String,
    val type: String,
    val action: String,
    val wide: Boolean = false
)

private val buttons = listOf(
    CalcButton("MC", "memory", "memory-clear"),
    CalcButton("MR", "memory", "memory-recall"),
    CalcButton("M+", "memory", "memory-add"),
    CalcButton("M\u2212", "memory", "memory-sub"),

    CalcButton("C", "clear", "clear"),
    CalcButton("\u232b", "clear", "backspace"),
    CalcButton("%", "operator", "percent"),
    CalcButton("\u00f7", "operator", "/"),

    CalcButton("7", "digit", "7"),
    CalcButton("8", "digit", "8"),
    CalcButton("9", "digit", "9"),
    CalcButton("\u00d7", "operator", "*"),

    CalcButton("4", "digit", "4"),
    CalcButton("5", "digit", "5"),
    CalcButton("6", "digit", "6"),
    CalcButton("\u2212", "operator", "-"),

    CalcButton("1", "digit", "1"),
    CalcButton("2", "digit", "2"),
    CalcButton("3", "digit", "3"),
    CalcButton("+", "operator", "+"),

    CalcButton("0", "digit", "0", wide = true),
    CalcButton(".", "digit", "."),
    CalcButton("=", "equals", "=")
)

@Composable
fun ButtonPanelView(
    onButton: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    LazyVerticalGrid(
        columns = GridCells.Fixed(4),
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(buttons) { btn ->
            Button(
                onClick = { onButton(btn.action) },
                colors = ButtonDefaults.buttonColors(
                    containerColor = backgroundColor(btn.type)
                ),
                shape = MaterialTheme.shapes.medium,
                modifier = if (btn.wide) {
                    Modifier
                        .height(56.dp)
                        .fillMaxWidth()
                } else {
                    Modifier
                        .aspectRatio(1f)
                        .height(56.dp)
                }
            ) {
                Text(
                    text = btn.label,
                    fontSize = if (btn.type == "memory") 14.sp else 18.sp,
                    fontWeight = FontWeight.Medium,
                    color = foregroundColor(btn.type)
                )
            }
        }
    }
}

private fun backgroundColor(type: String): Color {
    return when (type) {
        "digit" -> Color(0xFF1A1A3D)
        "operator" -> Color(0xFFE8455F)
        "equals" -> Color(0xFF109C57)
        "clear" -> Color(0xFF543482)
        "memory" -> Color(0xFF1A1A3D)
        else -> Color.Gray
    }
}

private fun foregroundColor(type: String): Color {
    return when (type) {
        "digit" -> Color(0xFFE5E5E5)
        "memory" -> Color(0xFF878DA9)
        else -> Color.White
    }
}
